ic-agent/src/agent/mod.rs

```diff
@@ -1868,6 +1841,90 @@ impl<'agent> IntoFuture for UpdateBuilder<'agent> {
     }
 }
 
+/// HTTP client middleware. Implemented automatically for `reqwest`-compatible by-ref `tower::Service`, such as `reqwest_middleware`.
+pub trait HttpService: Send + Sync {
+    /// Perform a HTTP request. Any retry logic should call `req` again, instead of `Request::try_clone`.
+    fn call<'a>(
+        &'a self,
+        req: &'a (dyn Fn() -> Result<Request, AgentError> + Send + Sync),
+        max_retries: usize,
+    ) -> AgentFuture<'a, Response>;
+}
+#[cfg(not(target_family = "wasm"))]
+impl<T> HttpService for T
+where
+    for<'a> &'a T: Service<Request, Response = Response, Error = reqwest::Error>,
+    for<'a> <&'a Self as Service<Request>>::Future: Send,
+    T: Send + Sync + ?Sized,
+{
+    fn call<'a>(
+        mut self: &'a Self,
+        req: &'a (dyn Fn() -> Result<Request, AgentError> + Send + Sync),
+        max_retries: usize,
+    ) -> AgentFuture<'a, Response> {
+        Box::pin(async move {
+            let mut retry_count = 0;
+            loop {
+                match Service::call(&mut self, req()?).await {
+                    Err(err) => {
+                        // Network-related errors can be retried.
+                        if err.is_connect() {
+                            if retry_count >= max_retries {
+                                return Err(AgentError::TransportError(err));
+                            }
+                            retry_count += 1;
+                        }
+                    }
+                    Ok(resp) => return Ok(resp),
+                }
+            }
+        })
+    }
+}
+
+#[cfg(target_family = "wasm")]
+impl<T> HttpService for T
+where
+    for<'a> &'a T: Service<Request, Response = Response, Error = reqwest::Error>,
+    T: Send + Sync + ?Sized,
+{
+    fn call<'a>(
+        mut self: &'a Self,
+        req: &'a (dyn Fn() -> Result<Request, AgentError> + Send + Sync),
+        _: usize,
+    ) -> AgentFuture<'a, Response> {
+        Box::pin(async move { Ok(Service::call(&mut self, req()?).await?) })
+    }
+}
+
+struct Retry429Logic {
+    client: Client,
+}
+
+impl HttpService for Retry429Logic {
+    fn call<'a>(
+        &'a self,
+        req: &'a (dyn Fn() -> Result<Request, AgentError> + Send + Sync),
+        _max_retries: usize,
+    ) -> AgentFuture<'a, Response> {
+        Box::pin(async move {
+            loop {
+                #[cfg(not(target_family = "wasm"))]
+                let resp = self.client.call(req, _max_retries).await?;
```

Comment:
Maybe add `max_retries` for 429 instead of retrying indefinitely?

Reply:
The reason for retrying 429s indefinitely is because you are not served them indefinitely - even though it means the semaphore isn't doing its job, your message backlog will eventually be cleared, or if it grows too much, then you'll start getting message expiry timeouts. Is this bad logic?